File: togo/internal/cli/audit_test.go
package cli

import (
	"strings"
	"testing"

	"github.com/nikeee/cappu/internal/audit"
	"github.com/nikeee/cappu/internal/packages"
)

func TestBuildAuditSarif(t *testing.T) {
	root := packages.NewCoordinates("org.apache.logging.log4j", "log4j-core", "2.14.1")
	byKey := map[packages.PackageKey]packages.ResolvedPackage{
		root.Key(): {Coordinates: root}, // zero RequestedBy => a declared root
	}
	report := audit.AuditReport{
		Scanned: 1,
		Vulnerable: []audit.PackageAdvisories{{
			Coordinates: root,
			Advisories: []audit.Advisory{
				{ID: "GHSA-jfh8-c2jp-5v3q", Aliases: []string{"CVE-2021-44228"}, Summary: "Log4Shell", Severity: audit.SeverityCritical, FixedVersions: []string{"2.15.0"}, URL: "https://example/ghsa"},
				{ID: "GHSA-minor", Summary: "minor issue", Severity: audit.SeverityLow},
			},
		}},
	}

	log := buildAuditSarif(report, byKey, "9.9.9")

	if log.Version != "2.1.0" || log.Schema == "" {
		t.Fatalf("expected SARIF 2.1.0 with schema, got %+v", log)
	}
	if len(log.Runs) != 1 {
		t.Fatalf("runs = %d, want 1", len(log.Runs))
	}
	run := log.Runs[0]
	if run.Tool.Driver.Name != "cappu" || run.Tool.Driver.Version != "9.9.9" {
		t.Errorf("driver = %+v", run.Tool.Driver)
	}
	if len(run.Tool.Driver.Rules) != 2 {
		t.Fatalf("rules = %d, want 2 (one per distinct advisory)", len(run.Tool.Driver.Rules))
	}
	scores := map[string]string{}
	for _, r := range run.Tool.Driver.Rules {
		scores[r.ID] = r.Properties.SecuritySeverity
	}
	if scores["GHSA-jfh8-c2jp-5v3q"] != "9.0" || scores["GHSA-minor"] != "1.0" {
		t.Errorf("security-severity scores = %v", scores)
	}

	if len(run.Results) != 2 {
		t.Fatalf("results = %d, want 2", len(run.Results))
	}
	var crit *sarifResult
	for i := range run.Results {
		if run.Results[i].RuleID == "GHSA-jfh8-c2jp-5v3q" {
			crit = &run.Results[i]
		}
	}
	if crit == nil {
		t.Fatal("missing critical result")
	}
	if crit.Level != "error" {
		t.Errorf("critical level = %q, want error", crit.Level)
	}
	if crit.Properties.Severity != "critical" {
		t.Errorf("severity property = %q", crit.Properties.Severity)
	}
	if len(crit.Locations) != 1 || crit.Locations[0].PhysicalLocation.ArtifactLocation.URI != "cappu.json" {
		t.Errorf("location = %+v", crit.Locations)
	}
	if n := len(crit.Properties.Path); n == 0 || crit.Properties.Path[n-1] != crit.Properties.Coordinate {
		t.Errorf("path should end at the vulnerable coordinate: %v", crit.Properties.Path)
	}
	for _, want := range []string{"CVE-2021-44228", "log4j-core:2.14.1", "Fixed in: 2.15.0."} {
		if !strings.Contains(crit.Message.Text, want) {
			t.Errorf("message %q missing %q", crit.Message.Text, want)
		}
	}
}

File: src/cli/audit.test.ts
import { test } from "node:test";

import { expect } from "expect";

import { type Advisory, type AuditReport } from "../audit/index.ts";
import {
  type Coordinates,
  packageKey,
  type ResolvedPackage,
  toCoordinates,
} from "../packages/index.ts";
import { buildAuditSarif } from "./audit.ts";

const coord = (spec: string): Coordinates => {
  const [g, a, v] = spec.split(":");
  return toCoordinates(g!, a!, v!);
};

const advisory = (a: Partial<Advisory> & Pick<Advisory, "id" | "severity">): Advisory => ({
  aliases: [],
  summary: "",
  fixedVersions: [],
  url: "",
  ...a,
});

test("buildAuditSarif emits a code-scanning SARIF 2.1.0 log", () => {
  const root = coord("org.app:app:1.0");
  const vuln = coord("org.apache.logging.log4j:log4j-core:2.14.1");
  // dependencyPath only reads coordinates/requestedBy, so stub the rest.
  const byKey = new Map<string, ResolvedPackage>([
    [packageKey(root), { coordinates: root } as ResolvedPackage],
    [packageKey(vuln), { coordinates: vuln, requestedBy: root } as ResolvedPackage],
  ]);
  const report: AuditReport = {
    scanned: 2,
    counts: { critical: 1, high: 0, moderate: 0, low: 1, unknown: 0 },
    vulnerable: [
      {
        coordinates: vuln,
        advisories: [
          advisory({
            id: "GHSA-jfh8-c2jp-5v3q" as Advisory["id"],
            aliases: ["CVE-2021-44228"],
            summary: "Log4Shell",
            severity: "critical",
            fixedVersions: ["2.15.0"],
            url: "https://example/ghsa",
          }),
          advisory({ id: "GHSA-minor" as Advisory["id"], summary: "minor issue", severity: "low" }),
        ],
      },
    ],
  };

  // buildAuditSarif returns `object`; narrow it for the assertions below.
  const sarif = buildAuditSarif(report, byKey, "9.9.9") as {
    $schema: string;
    version: string;
    runs: {
      tool: {
        driver: {
          name: string;
          version: string;
          rules: { id: string; properties: { "security-severity"?: string } }[];
        };
      };
      results: {
        ruleId: string;
        level: string;
        message: { text: string };
        locations: { physicalLocation: { artifactLocation: { uri: string } } }[];
        properties: { coordinate: string; severity: string; path: string[] };
      }[];
    }[];
  };

  expect(sarif.version).toBe("2.1.0");
  expect(sarif.$schema).toContain("sarif-2.1.0");
  expect(sarif.runs).toHaveLength(1);

  const driver = sarif.runs[0]!.tool.driver;
  expect(driver.name).toBe("cappu");
  expect(driver.version).toBe("9.9.9");
  expect(driver.rules).toHaveLength(2); // one rule per distinct advisory
  const scores = Object.fromEntries(
    driver.rules.map(r => [r.id, r.properties["security-severity"]]),
  );
  expect(scores["GHSA-jfh8-c2jp-5v3q"]).toBe("9.0");
  expect(scores["GHSA-minor"]).toBe("1.0");

  const results = sarif.runs[0]!.results;
  expect(results).toHaveLength(2);
  const crit = results.find(r => r.ruleId === "GHSA-jfh8-c2jp-5v3q")!;
  expect(crit.level).toBe("error");
  expect(crit.properties.severity).toBe("critical");
  expect(crit.locations[0]!.physicalLocation.artifactLocation.uri).toBe("cappu.json");
  expect(crit.properties.path).toEqual([
    "org.app:app:1.0",
    "org.apache.logging.log4j:log4j-core:2.14.1",
  ]);
  expect(crit.message.text).toContain("CVE-2021-44228");
  expect(crit.message.text).toContain("log4j-core:2.14.1");
  expect(crit.message.text).toContain("Fixed in: 2.15.0.");
});
